polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -2101,23 +2105,83 @@ ValueCategory MLIRScanner::VisitBinAssign(BinaryOperator *E) {
+            BinaryOperator::Opcode Opcode, FPOptions FPFeatures,
+            const Expr *Expr)
+      : LHS(LHS), RHS(RHS), Ty(Ty), Opcode(Opcode), FPFeatures(FPFeatures),
+        E(Expr) {}
 
   ValueCategory getLHS() const { return LHS; }
   ValueCategory getRHS() const { return RHS; }
   constexpr QualType getType() const { return Ty; }
   constexpr BinaryOperator::Opcode getOpcode() const { return Opcode; }
+  FPOptions getFPFeatures() const { return FPFeatures; }
   constexpr const Expr *getExpr() const { return E; }
 
 private:
   const ValueCategory LHS;
   const ValueCategory RHS;
   const QualType Ty;                   // Computation Type.
   const BinaryOperator::Opcode Opcode; // Opcode of BinOp to perform
+  FPOptions FPFeatures;
   const Expr *E;
 };
 
+// Check whether it would be legal to emit a `math.fma` operation to represent
+// op and if so, build the fmuladd.
+//
+// Checks that (a) the operation is fusable, and (b) -ffp-contract=on.
+static std::optional<ValueCategory> tryEmitFMulAdd(const BinOpInfo &Op,
+                                                   OpBuilder &Builder,
+                                                   Location Loc,
+                                                   bool IsSub = false) {
+  const BinaryOperator::Opcode Opcode = Op.getOpcode();
+
+  assert((Opcode == BO_Add || Opcode == BO_AddAssign || Opcode == BO_Sub ||
+          Opcode == BO_SubAssign) &&
+         "Only fadd/fsub can be the root of an fmuladd.");
+
+  // Check whether this op is marked as fusable and fusion is allowed.
+  if (OmitFPContract || !Op.getFPFeatures().allowFPContractWithinStatement())
+    return {};
+
+  // Peek through fneg to look for fmul. Make sure fneg has no users, and that
+  // it is the only use of its operand.
+  constexpr auto isNegOperand =
+      [](ValueCategory Val) -> std::pair<ValueCategory, bool> {
+    auto Op = Val.val.getDefiningOp<arith::NegFOp>();
+    if (Op && Val.val.use_empty()) {
+      Value Operand = Op.getOperand();
+      if (Operand.hasOneUse())
+        return {ValueCategory(Operand, /*IsReference=*/false), true};
+    }
+    return {Val, false};
+  };
+  auto [LHS, NegLHS] = isNegOperand(Op.getLHS());
+  auto [RHS, NegRHS] = isNegOperand(Op.getRHS());
+
+  // We have a potentially fusable op. Look for a mul on one of the operands.
+  // Also, make sure that the mul result isn't used directly. In that case,
+  // there's no point creating a muladd operation.
+  constexpr auto tryEmit = [](OpBuilder &Builder, Location Loc,
+                              ValueCategory Original, ValueCategory LHS,
+                              ValueCategory RHS, bool NegLHS, bool NegMul,
+                              bool NegAdd) -> std::optional<ValueCategory> {
+    auto LHSOp = LHS.val.getDefiningOp<arith::MulFOp>();
+    if (LHSOp && (LHS.val.use_empty() || NegLHS)) {
+      // If we looked through fneg, erase it.
+      if (NegLHS)
+        Original.val.getDefiningOp()->erase();
```

Comment:
Do we need to explicitly erase here or can we rely on the canonicalization to detect the than-dead `fneg`?

Reply:
Yes, I guess I can just do that :+1: